### DISCUSSION TITLE
On the Role of the Devs

### DISCUSSION THREAD

**Discussion by mOctave** [endless-sky/endless-sky#10477] [General]
After another day of hours-long game politics discussions, I've had enough of the debate over what devs/contributors are supposed to do vs not do. The most annoying part of all this is that it's clearly laid out in the development manifesto, among other places, and yet it _still_ isn't being followed.

Before I launch into my somewhat ranty list of ~~demands~~ suggestions, I would like to say that I genuinely appreciate the work of all the devs. I think that you have a really horrible job, and I'm sorry you ever got so dedicated to this game as to be shoehorned into your current role. That said, here's what I think your role should be (and yes, I'd love to discuss these points and see if we can get to a position where we agree on them). Some of these suggestions may be a little extreme, but I'm going to lay them on the table anyways.

### Recent Problems

There's been two big problems that I've been noticing lately, especially in Derpy's absence. I don't think these are new issues. Actually, I think both were issues when MZ was still around, but as the development culture has changed they've stuck around to rear their ugly heads.

**Problem 1: Direct Veto, Full Stop**

There's been multiple cases recently where a developer has reviewed a PR for the first time, found a problem with it, told the person who opened the PR everything that's wrong with it, and then disappeared. To be fair, the dev involved does begin with some meaningful discussion. But they still disappear before they can finish the discussion with the contributor. This is a problem. It means that there are now at least three PRs open that I'm aware of (and I'm sure more that I'm not aware of), where the whole PR is sitting in limbo waiting to hear back from one specific person who was very active for two days and hasn't been heard from since. That sucks for the PR authors, especially with a release date coming up fast. That sucks for the devs, because now they're perpetually in the middle of an argument until they can spare a few contiguous hours and get in the right head space to sort things out.

Now, this problem is different from a dev saying that a PR is unacceptable and closing it. That's also unpleasant, but necessary. 

**Problem 2: One Review, Back to Another Project**

Similarly, there's been a couple of cases where a specific dev has left a very good review and brought up some great points. Then, the PR author addresses the points, requests another review, and gets crickets in response, despite the fact that the dev is still active and often producing their own PRs. Sometimes, this lasts for a couple of weeks. Sometimes, it lasts for [four months straight](https://github.com/endless-sky/endless-sky/pull/8621). Actually, I don't think this is as bad now as it used to be: I believe there were PRs that used to sit around for literal years waiting for MZ to get to them. Still, it's a huge issue.

With both of these problems, the main issue isn't that the dev found a problem with the PR. That's bound to happen, and is a good thing for the project. What isn't good is that the devs then become impossible to reach. They don't respond to comments on the PR or outright review requests. Sometimes, they don't respond to Discord DMs, being asked by other devs to look into things (I presume), or annoyed PR authors pinging them on Discord with a parodies of Beatles songs (which, by the way, should absolutely become part of the review request process moving forwards).

I'd say this is a problem with our devs, but the fact is that these issues have been occurring for years with more than half of the active developers. At that point, we're either making a huge mistake choosing our devs, or we have an underlying culture issue. I believe that the issue is the second: that there's a disconnect between contributors and devs, and way too many expectations as to what devs are supposed to do that aren't being properly communicated.

So, it's time for what I believe to be
### Developer Responsibilities

In order of priority, I believe that these are the (absolutely horrible) responsibilities of a developer:
1. **Maintain Consistency in the Game:** This is the reason why we have devs in the first place. Devs should be able to tell when something fits the game or doesn't. They should be trusted to make policy decisions regarding whether or not something can be admitted to the game. Current state of a PR doesn't come into this, but rather whether or not the ideal solution suggested by a PR fits with the game. Because of this, in hindsight I admire actions like how Derpy closed https://github.com/endless-sky/endless-sky/pull/7958, even though it sucked in the moment. If anything, I think devs should be more willing to hit both the merge and close buttons: we can always reopen a PR later, or revert changes if absolutely necessary.
2. **Be Available:** To me, this is huge. A dev has the power to effectively shut down a PR until every other dev has heard their opinion on something. That power should never be used. If a dev reviews a PR, it should be expected that they will personally be available to provide another review within a week, and that they will make doing so a priority. If a dev comments on a PR in a manner that could lead to further discussion, there is similarly the expectation that they will be free for the next week or so to follow up if someone responds to their comments. If no one responds for a while, then it's okay if they're not available. But I don't think a dev should ever review or comment on a PR if they don't expect to be able to follow up soon after. If they have major concerns about a PR, they should share those concerns with another dev who they think will be available. If they don't respond for more than about a month after someone responds to them and asks for clarification, I would say that they should be pinged by the person requesting clarification, and if they still don't respond, they aren't committed enough to what they said to do more than mention it in passing, and other devs should go on with the merge procedure.
3. **Be Approachable:** A dev should never outright reject something, even if they find it personally offensive. Their role is to stand there and diffuse conflicts, making game development run smoothly. Even when closing a PR, they should give clear reasons that aren't just "I don't like it." They should ideally be on both GitHub and Discord, and respond positively to pings on both. They should have all or most of the skills of a community manager, have read Grey's conversation resources in full, and be prepared to apologize. They should never find themselves defending their actions, but let their work speak for itself. They should always accept the fact that they may be working with a socially awkward 12-year-old who is guaranteed to take personally everything they say. In short, the devs should be more socially adept than anyone else on the development side of ES.
4. **Stay Up-to-Date:** A dev should stay on top of most of what's going on in ES, at least things they might be expected to review. If they review a PR and find a problem with it, they should first read over the entire PR discussion before finishing their review. I don't necessarily expect a dev to be able to keep up-to-date with absolutely everything going on. In fact, I think it might be a good idea if devs used GitHub's default notification settings for the ES repo rather than watching everything, so they could get updates that were specifically aimed at them.
5. **Do Administrative Work:** The development manifesto lists the role of a dev as those of a reviewer with gatekeeping responsibilities. So, they should have that role, and work on it when they've worked through their backlog of other stuff to do. Ideally, devs would also have GitHub admin privileges, but considering how handing out admin privileges has worked in the past, I'm not sure that would both be doable and not result in World War Three.
6. **Add New Content:** The lowest priority of devs should be to add new content (under which heading I include code and artwork). There have been 300 content contributors to ES, but only about a dozen devs. Devs should leave content creation to others except for when they are caught up on all the fun administrative work that's been thrown their way. Yes, this sucks. But it also means that devs can do something to reduce the issue and PR backlogs rather than add to them.

I'm aware that these responsibilities are a bit extreme, but I also hope that by putting them out there we might be able to reach some sort of compromise that isn't the current 1-6-5-4-3-2 order of priorities that we have right now.

One thing that I think this could lead to is a more fine-grained division of roles, whereby some devs step down and instead are given a position as senior reviewer/content creators, still partaking in the weekly dev chats without having merge permissions. Those devs would then be free to influence the course of the game, but wouldn't have the same level of responsibility for points 2, 3, and 4 (and would have more responsibility for point 6).

### Ideal Review Process on a PR

Here's my idealized fantasy about how a PR review process works, and the dev's role in it. This is more the ideal review process on a PR opened by a longtime contributor than one for a first time contributor; we might want to be a bit nicer to first-timers.

1. PR is opened, in draft state, somewhat incomplete.
2. Within a day of the PR being opened, a dev looks at the PR. Without even reviewing the PR's actual content, the dev makes a policy judgment on whether or not the PR should be worked on. If the PR isn't appropriate for ES, that dev should close the PR. If the dev was wrong to do so, it can be reopened later. Still, **every dev on the dev team should be trusted to make a policy decision like this**. You don't give the President of the US the nuclear codes and then expect them not to use them when they deem necessary. Likewise, a dev should be able to close a PR if they think it doesn't fit the game, and they should do so as early as possible.
3. Having rubberstamped the PR in concept, the devs should now ignore it. Reviewers should review it and get it to a near merge-ready state.
4. With the PR now being ready for merge in the eyes of the reviewers, the devs should return. The devs should point out any kinks they feel need to be worked out. At least two devs should probably be present in this stage unless the PR is some sort of small content PR (typo fix, changing weapon attribute, etc.).
5. A dev focused on the area of the PR (or possibly Derpy if no dev feels competent enough to judge) should merge the PR.

**Comment by roadrunner56** [comment]
So disclaimer that this is my own opinion and not necessarily that of the entire dev team:
I agree with most of the points here. I know that you mostly work on the code side of things, so I can't speak to that, but for the writing side of things, I know that responsibility 2 isn't always followed (I'm definitely not an exception to this). 
However, if I only ever review things where I was 100% sure I would be able to provide another comprehensive review within a week, I'd probably only ever do reviews during the summer. As it stands right now, I post a review, and if I don't return to explicitly approve the PR, assuming my comments have been judged by other devs to have been adequately addressed then they should feel free to merge the PR. I assume the other content devs feel the same way.
As for the review process, right now the only consistently active writing reviewer is Bene. This means that devs that focus on content and writing have to take on step 3 in the review process as well, as it's not fair to just dump every single writing PR on one person. I can't recall any situations where I saw a writing PR that was wildly inappropriate for vanilla ES - most large writing PRs (factions, new storylines) get run by other people on the Discord server, so by the time it gets PRed it's usually pretty well-thought out. However, yeah I do agree we should probably be better with doing conceptual reviews of large content PRs sooner rather than later. 
As for leaving content creation to others, it's a bit of a meme at this point that no one's ever going to finish the Wanderers/Republic campaigns, just because they're so big and require a lot of coordination with other parts of the game. With stuff of this scale that's been sitting unfinished for years, it isn't really feasible to keep waiting around until one guy writes the whole thing for us. Our current plans (keep in mind that this could change) are to have Doom make a PR for the Wanderers finale, and for the dev team to hash out a concrete outline for the entire Republic campaign, releasing this outline so that various people can write missions for it. This would get the ball rolling on the campaign and give it some overall coordination, without it being entirely on the dev team to finish the thing. So while most of the time I would agree that devs should be focusing on other people's content, I do feel there are exceptions to this.
Otherwise, you bring up some good points. I have a limited amount of time to spend on ES, so I've become increasingly aware of the importance of choosing how to spend that time.
